File: flutter_theoplayer_sdk_android/android/src/main/kotlin/com/theoplayer/flutter/AudioTrackBridge.kt
package com.theoplayer.flutter

import com.theoplayer.android.api.event.EventListener
import com.theoplayer.android.api.event.track.mediatrack.audio.ActiveQualityChangedEvent
import com.theoplayer.android.api.event.track.mediatrack.audio.AudioTrackEventTypes
import com.theoplayer.android.api.event.track.mediatrack.audio.TargetQualityChangedEvent
import com.theoplayer.android.api.event.track.mediatrack.audio.list.AddTrackEvent
import com.theoplayer.android.api.event.track.mediatrack.audio.list.AudioTrackListEventTypes
import com.theoplayer.android.api.event.track.mediatrack.audio.list.RemoveTrackEvent
import com.theoplayer.android.api.event.track.mediatrack.audio.list.TrackListChangeEvent
import com.theoplayer.android.api.event.track.mediatrack.quality.QualityEventTypes
import com.theoplayer.android.api.event.track.mediatrack.quality.QualityUpdateEvent
import com.theoplayer.android.api.player.Player
import com.theoplayer.android.api.player.track.mediatrack.MediaTrack
import com.theoplayer.android.api.player.track.mediatrack.quality.AudioQuality
import com.theoplayer.flutter.pigeon.THEOplayerFlutterAudioTracksAPI
import com.theoplayer.flutter.pigeon.THEOplayerNativeAudioTracksAPI
import com.theoplayer.flutter.pigeon.THEOplayerNativeAudioTracksAPI.Companion.setUp

class AudioTrackBridge(
    private val player: Player,
    pigeonMessenger: PigeonBinaryMessengerWrapper,
) : THEOplayerNativeAudioTracksAPI {

    private val flutterAudioTracksAPI = THEOplayerFlutterAudioTracksAPI(pigeonMessenger)

    init {
        setUp(pigeonMessenger, this)
    }

    private val emptyCallback: (Result<Unit>) -> Unit = {}

    private val addAudioTrackListener: EventListener<AddTrackEvent> = EventListener {
        flutterAudioTracksAPI.onAddAudioTrack(
            it.track.id,
            it.track.uid.toLong(),
            it.track.label,
            it.track.language,
            it.track.kind,
            it.track.isEnabled,
            emptyCallback
        )

        it.track.qualities?.forEach { quality ->
            flutterAudioTracksAPI.onAudioTrackAddQuality(
                it.track.uid.toLong(),
                quality.id,
                quality.uid.toLong(),
                quality.name,
                quality.bandwidth,
                quality.codecs,
                quality.audioSamplingRate,
                emptyCallback
            )
        }

        attachTrackListeners(it.track)
    }

    private val removeAudioTrackListener = EventListener<RemoveTrackEvent> {
        removeTrackListeners(it.track)
        flutterAudioTracksAPI.onRemoveAudioTrack(it.track.uid.toLong(), emptyCallback)
    }

    private val audioTrackListChangeListener = EventListener<TrackListChangeEvent> {
        flutterAudioTracksAPI.onAudioTrackListChange(it.track.uid.toLong(), emptyCallback)
    }

    private fun audioTrackTargetQualityChangedListener(track: MediaTrack<AudioQuality>) = EventListener<TargetQualityChangedEvent> {
        flutterAudioTracksAPI.onTargetQualityChange(
            track.uid.toLong(),
            it.qualities.map { audioQuality -> audioQuality.uid.toLong() },
            it.quality?.uid?.toLong(),
            emptyCallback
        )
    }

    private fun audioTrackActiveQualityChangedListener(track: MediaTrack<AudioQuality>) = EventListener<ActiveQualityChangedEvent> {
        flutterAudioTracksAPI.onActiveQualityChange(track.uid.toLong(), it.quality.uid.toLong(), emptyCallback)
    }

    private fun audioQualityChangeListener(track: MediaTrack<AudioQuality>, quality: AudioQuality) = EventListener<QualityUpdateEvent> {
        flutterAudioTracksAPI.onQualityUpdate(
            track.uid.toLong(),
            quality.uid.toLong(),
            quality.name,
            quality.bandwidth,
            quality.codecs,
            quality.audioSamplingRate,
            emptyCallback
        )
    }

    fun attachListeners() {
        player.audioTracks.addEventListener(AudioTrackListEventTypes.ADDTRACK, addAudioTrackListener)
        player.audioTracks.addEventListener(AudioTrackListEventTypes.REMOVETRACK, removeAudioTrackListener)
        player.audioTracks.addEventListener(AudioTrackListEventTypes.TRACKLISTCHANGE, audioTrackListChangeListener)
    }

    fun removeListeners() {
        player.audioTracks.removeEventListener(AudioTrackListEventTypes.ADDTRACK, addAudioTrackListener)
        player.audioTracks.removeEventListener(AudioTrackListEventTypes.REMOVETRACK, removeAudioTrackListener)
        player.audioTracks.removeEventListener(AudioTrackListEventTypes.TRACKLISTCHANGE, audioTrackListChangeListener)
        player.audioTracks.forEach { track ->
            removeTrackListeners(track)
        }
    }

    private fun attachTrackListeners(track: MediaTrack<AudioQuality>) {
        track.addEventListener(AudioTrackEventTypes.TARGETQUALITYCHANGEDEVENT, audioTrackTargetQualityChangedListener(track))
        track.addEventListener(AudioTrackEventTypes.ACTIVEQUALITYCHANGEDEVENT, audioTrackActiveQualityChangedListener(track))
        track.qualities?.forEach {
            attachQualityListeners(track, it)
        }
    }

    private fun removeTrackListeners(track: MediaTrack<AudioQuality>) {
        track.removeEventListener(AudioTrackEventTypes.TARGETQUALITYCHANGEDEVENT, audioTrackTargetQualityChangedListener(track))
        track.removeEventListener(AudioTrackEventTypes.ACTIVEQUALITYCHANGEDEVENT, audioTrackActiveQualityChangedListener(track))
        track.qualities?.forEach {
            removeQualityListeners(track, it)
        }
    }

    private fun attachQualityListeners(track: MediaTrack<AudioQuality>, quality: AudioQuality) {
        quality.addEventListener(QualityEventTypes.UPDATE, audioQualityChangeListener(track, quality))
    }

    private fun removeQualityListeners(track: MediaTrack<AudioQuality>, quality: AudioQuality) {
        quality.removeEventListener(QualityEventTypes.UPDATE, audioQualityChangeListener(track, quality))
    }

    override fun setTargetQuality(audioTrackUid: Long, qualityUid: Long?) {
        val audioTrack = player.audioTracks.find { it.uid.toLong() == audioTrackUid }
        val audioQuality = audioTrack?.qualities?.find { it.uid.toLong() == qualityUid }
        audioTrack?.targetQuality = audioQuality
    }

    override fun setTargetQualities(audioTrackUid: Long, qualitiesUid: List<Long>?) {
        val audioTrack = player.audioTracks.find { it.uid.toLong() == audioTrackUid }
        val audioQualities = audioTrack?.qualities?.filter { qualitiesUid?.contains(it.uid.toLong()) == true }
        audioTrack?.setTargetQualities(audioQualities)
    }

    override fun setEnabled(audioTrackUid: Long, enabled: Boolean) {
        val audioTrack = player.audioTracks.find { it.uid.toLong() == audioTrackUid }
        audioTrack?.isEnabled = enabled
    }

}

File: flutter_theoplayer_sdk/flutter_theoplayer_sdk_ios/ios/Classes/AudioTrackBridge.swift
//
//  AudioTrackBridge.swift
//  flutter_theoplayer_sdk_ios
//
//  Created by Hovig on 27/11/2023.
//

import Foundation
import THEOplayerSDK
import Flutter

class AudioTrackBridge: THEOplayerNativeAudioTracksAPI {

    private let theoplayer: THEOplayer
    private let pigeonMessenger: PigeonBinaryMessengerWrapper
    private let flutterAudioTracksAPI: THEOplayerFlutterAudioTracksAPI

    private var addAudioTrackListener: EventListener?
    private var removeAudioTrackListener: EventListener?
    private var audioTrackListChangeListener: EventListener?
    private var trackEventListeners: [Int: [EventListener]] = [:]

    private let emptyCompletion: (Result<Void, PigeonError>) -> Void = {result in }

    init(theoplayer: THEOplayer, pigeonMessenger: PigeonBinaryMessengerWrapper) {
        self.theoplayer = theoplayer
        self.pigeonMessenger = pigeonMessenger
        self.flutterAudioTracksAPI = THEOplayerFlutterAudioTracksAPI(binaryMessenger: pigeonMessenger)
        THEOplayerNativeAudioTracksAPISetup.setUp(binaryMessenger: pigeonMessenger, api: self)
    }

    func attachListeners() {
        addAudioTrackListener = theoplayer.audioTracks.addEventListener(type: AudioTrackListEventTypes.ADD_TRACK, listener: { [weak self] event in
            guard let self else { return }
            if let audioTrack = event.track as? AudioTrack {
                self.flutterAudioTracksAPI.onAddAudioTrack(
                    id: audioTrack.id,
                    uid: Int64(audioTrack.uid),
                    label: audioTrack.label,
                    language: audioTrack.language,
                    kind: audioTrack.kind,
                    isEnabled: audioTrack.enabled,
                    unlocalizedLabel: audioTrack.unlocalizedLabel,
                    completion: self.emptyCompletion)

                // Emit quality info for each quality in the track
                let qualities = audioTrack.qualities
                for i in 0..<qualities.count {
                    let quality = qualities.get(i)
                    self.flutterAudioTracksAPI.onAudioTrackAddQuality(
                        audioTrackUid: Int64(audioTrack.uid),
                        qualityId: String(quality.bandwidth),
                        qualityUid: Int64(quality.bandwidth),
                        name: self.labelFromBandwidth(quality.bandwidth),
                        bandwidth: Int64(quality.bandwidth),
                        codecs: "",
                        audioSamplingRate: 0,
                        averageBandwidth: quality.averageBandwidth.map { Int64($0) },
                        available: true,
                        completion: self.emptyCompletion)
                }

                self.attachTrackListeners(audioTrack: audioTrack)
            }
        })

        removeAudioTrackListener = theoplayer.audioTracks.addEventListener(type: AudioTrackListEventTypes.REMOVE_TRACK, listener: { [weak self] event in
            guard let self else { return }
            self.removeTrackListeners(trackUid: event.track.uid)
            self.flutterAudioTracksAPI.onRemoveAudioTrack(uid: Int64(event.track.uid), completion: self.emptyCompletion)
        })

        audioTrackListChangeListener = theoplayer.audioTracks.addEventListener(type: AudioTrackListEventTypes.CHANGE, listener: { [weak self] event in
            guard let self else { return }
            self.flutterAudioTracksAPI.onAudioTrackListChange(uid: Int64(event.track.uid), completion: self.emptyCompletion)
        })
    }

    private func attachTrackListeners(audioTrack: AudioTrack) {
        var listeners: [EventListener] = []

        let activeQualityChangedListener = audioTrack.addEventListener(type: MediaTrackEventTypes.ACTIVE_QUALITY_CHANGED, listener: { [weak self] (event: ActiveQualityChangedEvent) in
            guard let self else { return }
            self.flutterAudioTracksAPI.onActiveQualityChange(
                audioTrackUid: Int64(audioTrack.uid),
                qualityUid: Int64(event.quality.bandwidth),
                completion: self.emptyCompletion)
        })
        listeners.append(activeQualityChangedListener)

        trackEventListeners[audioTrack.uid] = listeners
    }

    private func removeTrackListeners(trackUid: Int) {
        trackEventListeners.removeValue(forKey: trackUid)
    }

    private func removeListeners() {
        if let listener = addAudioTrackListener {
            theoplayer.audioTracks.removeEventListener(type: AudioTrackListEventTypes.ADD_TRACK, listener: listener)
        }
        if let listener = removeAudioTrackListener {
            theoplayer.audioTracks.removeEventListener(type: AudioTrackListEventTypes.REMOVE_TRACK, listener: listener)
        }
        if let listener = audioTrackListChangeListener {
            theoplayer.audioTracks.removeEventListener(type: AudioTrackListEventTypes.CHANGE, listener: listener)
        }
        trackEventListeners.removeAll()
    }

    func dispose() {
        removeListeners()
        THEOplayerNativeAudioTracksAPISetup.setUp(binaryMessenger: pigeonMessenger, api: nil)
    }

    func setTargetQuality(audioTrackUid: Int64, qualityUid: Int64?) throws {
        for i in 0..<theoplayer.audioTracks.count {
            let track = theoplayer.audioTracks.get(i)
            if track.uid == Int(audioTrackUid) {
                if let qualityUid = qualityUid, let qualities = (track as? AudioTrack)?.qualities {
                    for j in 0..<qualities.count {
                        let quality = qualities.get(j)
                        if quality.bandwidth == Int(qualityUid) {
                            (track as? AudioTrack)?.targetQualities = [quality]
                            return
                        }
                    }
                } else {
                    (track as? AudioTrack)?.targetQualities = nil
                }
                return
            }
        }
    }

    func setTargetQualities(audioTrackUid: Int64, qualitiesUid: [Int64]?) throws {
        for i in 0..<theoplayer.audioTracks.count {
            let track = theoplayer.audioTracks.get(i)
            if track.uid == Int(audioTrackUid) {
                if let qualitiesUid = qualitiesUid, let qualities = (track as? AudioTrack)?.qualities {
                    var targetQualities: [any Quality] = []
                    for j in 0..<qualities.count {
                        let quality = qualities.get(j)
                        if qualitiesUid.contains(Int64(quality.bandwidth)) {
                            targetQualities.append(quality)
                        }
                    }
                    (track as? AudioTrack)?.targetQualities = targetQualities
                } else {
                    (track as? AudioTrack)?.targetQualities = nil
                }
                return
            }
        }
    }

    func setEnabled(audioTrackUid: Int64, enabled: Bool) throws {
        for i in 0..<theoplayer.audioTracks.count {
            var track = theoplayer.audioTracks.get(i)
            if (track.uid == audioTrackUid) {
                track.enabled = enabled
            }
        }
    }

    private func labelFromBandwidth(_ bandwidth: Int) -> String {
        if bandwidth >= 1_000_000 {
            return String(format: "%.1f Mbps", Double(bandwidth) / 1_000_000.0)
        } else {
            return String(format: "%d kbps", bandwidth / 1000)
        }
    }

}
